### DISCUSSION TITLE
app.tsx中可以使用自定义hook获取数据吗？

### DISCUSSION THREAD

**Discussion by Chill7773** [umijs/umi#11885] [Q&A] [answered]
有个公共的自定义hook，需要在app.tsx中获取其值，并赋值到layout的参数中。需要怎么操作？

**Comment by fz6m** [comment]
```ts
// src/app.tsx

let config;

export async function getInitialState() {
  config = await api()
  return ...
}

export const layout = () => {
  // use config
}
```

**Reply by Chill7773** [reply]
不一样的是，我的函数是个react hook，无法在 getInitialState中调用。app.tsx有其他地方可以使用react hook吗？

**Reply by fz6m** [reply]
layout 的配置被使用的时机是 pro layout 渲染时，假如你获取数据的时机晚于 layout 渲染是不行的，较好的时机是在应用启动时获取到然后设定好 layout ，最晚的设定时机不应该晚于界面的 layout 渲染时。

比较好的方法如上文回复中所示，使用 `getInitialState` 在其中调用 api 获取，另外可以考虑在 `innerProvider` 的上下文层级获取 ：

```ts
// src/app.tsx

export const innerProvider = (container) => {
  return <Context>{container}</Context>
}

function Context({ children }) {
  
  // 此处考虑调用 api 获取到数据后存放到 valtio 等可以脱离 hooks 的全局数据管理方案内，之后在 src/app.tsx 中 export const layout 内使用即可

  if (!hasData) {
    return null // or loading
  }
  return children
}
```

**Reply by Chill7773** [reply]
innerProvider是新的api吗？我在umi的官网上没有找到有关innerProvider的介绍

**Reply by fz6m** [reply, marked as answer]
不是，umi 里面 context 有很多层级：

```ts
const UMI_CLIENT_RENDER_REACT_PLUGIN_LIST = [
  // Lowest to the highest priority
  'innerProvider',
  'i18nProvider',
  'accessProvider',
  'dataflowProvider',
  'outerProvider',
  'rootContainer',
];
```

通常我们使用 `rootContainer` 去包我们自己的 context 是没任何问题的，但如果你需要在 context 里用到 model 、i18n locale 等方法的时候，就需要 context 的层级比较低，`innerProvider` 是离应用渲染内容最近的一个层级，一般情况用不到。

**Reply by Chill7773** [reply]
多谢！有所帮助！
